src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/RichTextEditorTextArea.kt

```diff
@@ -0,0 +1,65 @@
+package com.atlassian.performance.tools.jiraactions.api.page
+
+import com.atlassian.performance.tools.jiraactions.Patience
+import org.apache.logging.log4j.LogManager
+import org.apache.logging.log4j.Logger
+import org.openqa.selenium.By
+import org.openqa.selenium.JavascriptExecutor
+import org.openqa.selenium.WebDriver
+import org.openqa.selenium.WebElement
+import org.openqa.selenium.support.ui.ExpectedConditions
+import org.openqa.selenium.support.ui.ExpectedConditions.*
+import java.time.Duration
+import org.openqa.selenium.support.ui.WebDriverWait
+
+
+
+/**
+ * Helper class for Rich Text Editor enabled Textarea
+ *
+ */
+class RichTextEditorTextArea(
+    private val driver: WebDriver
+) {
+
+    private val logger: Logger = LogManager.getLogger(this::class.java)
+
+    fun overwriteIfPresent(id : String, text: String, optional : Boolean = false) {
+
+        //for optional field, ignore it if not found
+        if (optional && !driver.isElementPresent(By.id(id))) return
+
+        val wait = WebDriverWait(driver, 2)
+        val webElement = wait.until { driver -> driver.findElement(By.id(id)) }
+        val attrClass = webElement.getAttribute("class")
+        if(attrClass != null && attrClass.contains(other = "richeditor-cover", ignoreCase = true)){
+            overwriteRich(id, text)
+        }
+        else{
+            //try plain text editing
+            overwritePlain(webElement, text)
+        }
+    }
+
+    private fun overwritePlain(webElement: WebElement, text: String) {
+        driver.wait(timeout = Duration.ofSeconds(2), condition = elementToBeClickable(webElement))
+        webElement.clear()
+        webElement.sendKeys(text)
+    }
+
+    private fun overwriteRich(id: String, text: String) {
+        val iframeXpath = "//div[textarea[@id='$id']]//iframe"
+        //find the tinymce iFrame
+        driver.wait(timeout = Duration.ofSeconds(10), condition = frameToBeAvailableAndSwitchToIt(By.xpath(iframeXpath)))
+
+        //tinymce uses body element with id 'tinymce' to hold the text, find the body element
+        val bodyElement = driver.findElement(By.id("tinymce"))
+        //data-projectkey={project}
+        driver.wait(timeout = Duration.ofSeconds(10), condition = attributeToBeNotEmpty(bodyElement, "data-projectkey"))
+        bodyElement.clear()
+        bodyElement.sendKeys(text)
+
+        driver.switchTo().parentFrame()
```

Comment:
If there's an exception before this line, we will not return to the parent frame.

```kotlin
try {
    driver.wait(
        condition = frameToBeAvailableAndSwitchToIt(By.xpath(iframeXpath)),
        timeout = Duration.ofSeconds(10)
    )
    val tinyMce = driver.findElement(By.id("tinymce"))
    driver.wait(
        condition = attributeToBeNotEmpty(tinyMce, "data-projectkey"),
        timeout = Duration.ofSeconds(10)
    )
    tinyMce.clear()
    tinyMce.sendKeys(text)
} finally {
    driver.switchTo().parentFrame()
}
```
This makes sure we always switch to the parent frame, regardless of exceptions within the `try`.
Not that I also removed the inline comments, because they said literally the same words as the code already did.

Reply:
good point, will do